File: cal-reducer/src/state.ts
/**
 * Protocol State model + genesis + STATE_ROOT (CAL Spec §7.3, reducer design §3).
 *
 * State is the canonical-JSON map of the eight namespaces. Internally we key it
 * by the short namespace name (`cal`, `ptra`, …); the §7.3 leaf uses the full
 * `state.<name>` form, applied in `stateRootOf`. Integers are bigint; STATE_ROOT
 * is computed by the canonical layer (already NORMATIVE), so the reducer never
 * reimplements §7.3.
 */

import { canonicalizeValue, stateRoot } from "@paradigm-terra/canonical";

/** JSON value as produced/consumed by the canonical layer. */
export type Json = null | boolean | bigint | string | Json[] | { [k: string]: Json };

/** The eight namespaces, in the §7.3 UTF-8-sorted order (by `state.<name>`). */
export const NAMESPACES = [
  "cal",
  "failure_mode",
  "governance",
  "oracles",
  "ptra",
  "registry",
  "tick",
  "treasury",
] as const;

export type State = { [k: string]: Json };

/** The fixed genesis state. Its STATE_ROOT is pinned by the golden vectors. */
export function genesis(): State {
  return {
    cal: { in_flight: {}, nonces: {} },
    failure_mode: { is_bounded_mode: false, capture_guard_counters: {} },
    governance: { gas_price_nano_ptra_per_unit: 1000n, genesis_validator_set: [], params: {} },
    oracles: { feeds: {} },
    ptra: { balances: {} },
    registry: { agents: {}, mcp_schema_hash: `0x${"00".repeat(32)}` },
    tick: { current: 0n },
    treasury: { nav: 0n, developer_fund_balance: 0n, collected_fees_window: 0n },
  };
}

/** STATE_ROOT over the eight namespaces (CAL Spec §7.3). */
export function stateRootOf(state: State): Uint8Array {
  return stateRoot(
    NAMESPACES.map((n) => ({ name: `state.${n}`, canonicalBytes: canonicalizeValue(state[n]) })),
  );
}

function isPlainObject(v: Json | undefined): v is { [k: string]: Json } {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

/** Read a value at a dotted path; `undefined` if any segment is missing. */
export function getIn(obj: Json | undefined, path: readonly string[]): Json | undefined {
  let cur: Json | undefined = obj;
  for (const seg of path) {
    if (!isPlainObject(cur) || !(seg in cur)) return undefined;
    cur = cur[seg];
  }
  return cur;
}

/** Immutably set a value at a path, cloning only along the path. */
export function setIn(obj: Json | undefined, path: readonly string[], value: Json): Json {
  if (path.length === 0) return value;
  const [head, ...rest] = path as [string, ...string[]];
  const base = isPlainObject(obj) ? obj : {};
  return { ...base, [head]: setIn(base[head], rest, value) };
}

/** Immutably delete the key at a path (no-op if absent). */
export function deleteIn(obj: Json | undefined, path: readonly string[]): Json {
  if (!isPlainObject(obj)) return obj ?? {};
  if (path.length === 1) {
    const { [path[0]!]: _drop, ...rest } = obj;
    return rest;
  }
  const [head, ...rest] = path as [string, ...string[]];
  if (!(head in obj)) return obj;
  return { ...obj, [head]: deleteIn(obj[head], rest) };
}

File: cal-reducer-go/state.go
package calreducer

import (
	"strings"

	canonical "github.com/paradigm-terra/canonical-go"
)

// Namespaces in the §7.3 UTF-8-sorted order (by `state.<name>`).
var Namespaces = []string{"cal", "failure_mode", "governance", "oracles", "ptra", "registry", "tick", "treasury"}

// Genesis is the fixed genesis state (its STATE_ROOT is pinned by the golden vectors).
func Genesis() canonical.Value {
	o, p, iu, a := canonical.O, canonical.P, canonical.IntU, canonical.A
	return o(
		p("cal", o(p("in_flight", o()), p("nonces", o()))),
		p("failure_mode", o(p("is_bounded_mode", false), p("capture_guard_counters", o()))),
		p("governance", o(p("gas_price_nano_ptra_per_unit", iu(1000)), p("genesis_validator_set", a()), p("params", o()))),
		p("oracles", o(p("feeds", o()))),
		p("ptra", o(p("balances", o()))),
		p("registry", o(p("agents", o()), p("mcp_schema_hash", "0x"+strings.Repeat("00", 32)))),
		p("tick", o(p("current", iu(0)))),
		p("treasury", o(p("nav", iu(0)), p("developer_fund_balance", iu(0)), p("collected_fees_window", iu(0)))),
	)
}

// StateRootOf computes the STATE_ROOT over the eight namespaces (CAL Spec §7.3).
func StateRootOf(state canonical.Value) ([32]byte, error) {
	nss := make([]canonical.StateNamespace, 0, len(Namespaces))
	for _, n := range Namespaces {
		content, _ := getIn(state, []string{n})
		b, err := canonical.CanonicalizeValue(content)
		if err != nil {
			return [32]byte{}, err
		}
		nss = append(nss, canonical.StateNamespace{Name: "state." + n, CanonicalBytes: b})
	}
	return canonical.StateRoot(nss)
}

func getIn(v canonical.Value, path []string) (canonical.Value, bool) {
	cur := v
	for _, seg := range path {
		o, ok := cur.(*canonical.Object)
		if !ok {
			return nil, false
		}
		cv, ok := o.Get(seg)
		if !ok {
			return nil, false
		}
		cur = cv
	}
	return cur, true
}

func setIn(v canonical.Value, path []string, newval canonical.Value) canonical.Value {
	if len(path) == 0 {
		return newval
	}
	head := path[0]
	o, _ := v.(*canonical.Object)
	var pairs []canonical.Pair
	found := false
	if o != nil {
		for _, k := range o.Keys() {
			cv, _ := o.Get(k)
			if k == head {
				pairs = append(pairs, canonical.P(head, setIn(cv, path[1:], newval)))
				found = true
			} else {
				pairs = append(pairs, canonical.P(k, cv))
			}
		}
	}
	if !found {
		pairs = append(pairs, canonical.P(head, setIn(nil, path[1:], newval)))
	}
	return canonical.NewObject(pairs...)
}

func deleteIn(v canonical.Value, path []string) canonical.Value {
	o, ok := v.(*canonical.Object)
	if !ok {
		return v
	}
	var pairs []canonical.Pair
	for _, k := range o.Keys() {
		cv, _ := o.Get(k)
		if len(path) == 1 {
			if k == path[0] {
				continue
			}
			pairs = append(pairs, canonical.P(k, cv))
		} else if k == path[0] {
			pairs = append(pairs, canonical.P(k, deleteIn(cv, path[1:])))
		} else {
			pairs = append(pairs, canonical.P(k, cv))
		}
	}
	return canonical.NewObject(pairs...)
}
